### ISSUE TITLE
Prevent certain events from triggering the same turn

### ISSUE THREAD

**Issue by Nightinggale** [We-the-People-civ4col-mod/Mod#901]
Right now event triggering is done as following:
All event triggers are looped. For those, which can trigger, all with negative weight triggers instantly. This is followed by all with positive weight where a single one is picked randomly according to weight.

The problem is the ones with negative weight as it allows triggering multiple events at once, which all cost money and the player can only afford one. This then results in the rest being lost and in some cases can never trigger again.

Solution 1:
Add a list of event triggers to each event trigger, which states that the event trigger can't trigger on the same turn as any of the listed event triggers.

Solution 2:
Add an answer to certain events, which rejects, but also resets that the event triggered. This in turn will make a one time event able to trigger again later.

Nothing prevents us from implementing both solutions.

-----

CvPlayer::doEvents needs improvements in general. Switch to looping enums, const correctness, get rid of getDefineINT etc.

There is a hidden performance killer. CvPlayer::isTriggerFired keeps track of which events have been triggered at least once. This is done by adding the enum to a vector. With more than 500 event triggers, this list can become huge. Since it's in random order, each time an event is checked, the entire list has to be looped to verify that it's not present. This scales poorly and eats up a bunch of memory.

Replacing it with an EnumMap storing bools will prevent the need for looping. Also it reduces memory usage by storing 32 bools in the same space as we currently stores a single enum entry. Setting or clearing will also become much easier than with the current code.

**Comment by devolution79** [comment]
Another note on performance: Be careful with implementing trigger checks in python since these can be very slow due to the overhead of calling into python and the execution speed of python code in general. Also, there is a lot of duplicated checks for valid event pre-conditions in the python code and these could certainly be refactored (and like previously noted, try to make them general and move them into the DLL)

**Comment by devolution79** [comment]
I wonder if we can just let ChatGPT loose on the python event code and let it do it's thing :stuck_out_tongue:

**Comment by devolution79** [comment]
I realize now that this was a bit off-topic so we may want to move the performance stuff to a separate issue

**Comment by Nightinggale** [comment]
> Another note on performance: Be careful with implementing trigger checks in python since these can be very slow due to the overhead of calling into python and the execution speed of python code in general.

Agreed. We should ideally do the checks in C++, through in general performance is off topic for this issue. My plan is to have a temp EnumMap to store what triggered this turn. This would be high performance, particularly since it will remain unallocated until something with negative weight triggers, meaning most turns it will not even be allocated.
